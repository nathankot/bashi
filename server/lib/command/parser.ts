import * as t from "io-ts";
import * as p from "typescript-parsec";

import {
  // Token,
  buildLexer,
  rule,
  expectEOF,
  expectSingleResult,
} from "typescript-parsec";

import { Value } from "@lib/valueTypes.ts";

interface Call {
  type: "call";
  name: string;
  args: (Value | Call)[];
}

const Call: t.Type<Call> = t.recursion("Call", () =>
  t.type({
    type: t.literal("call"),
    name: t.string,
    args: t.array(t.union([Value, Call])),
  })
);

export const ActionGroup = t.intersection([
  t.type({
    thought: t.string,
    action: t.string,
    functionCalls: t.array(Call),
  }),
  t.partial({
    result: t.string,
  }),
]);
export type ActionGroup = t.TypeOf<typeof ActionGroup>;

enum ActionTokenKind {
  KeywordThought,
  KeywordAction,
  KeywordResult,

  String,

  Newline,
  Space,
}

const actionLexer = buildLexer([
  [true, /^Action/gi, ActionTokenKind.KeywordAction],
  [true, /^Thought/gi, ActionTokenKind.KeywordThought],
  [true, /^Result/gi, ActionTokenKind.KeywordResult],

  [true, /^:.*/g, ActionTokenKind.String],

  [true, /^(\n|\r\n|\r)+/g, ActionTokenKind.Newline],
  [false, /^\s+/g, ActionTokenKind.Space],
]);

const ACTION_GROUP = rule<ActionTokenKind, ActionGroup>();
ACTION_GROUP.setPattern(
  p.apply(
    p.seq(
      p.apply(
        p.kright(
          p.tok(ActionTokenKind.KeywordThought),
          p.tok(ActionTokenKind.String)
        ),
        (str) => str.text.slice(1).trim()
      ),
      p.tok(ActionTokenKind.Newline),
      p.apply(
        p.kright(
          p.tok(ActionTokenKind.KeywordAction),
          p.tok(ActionTokenKind.String)
        ),
        (str) => str.text.slice(1).trim()
      ),
      p.opt_sc(
        p.apply(
          p.kright(
            p.tok(ActionTokenKind.Newline),
            p.kright(
              p.tok(ActionTokenKind.KeywordResult),
              p.tok(ActionTokenKind.String)
            )
          ),
          (str) => {
            const result = str.text.slice(1).trim();
            if (result.length === 0) {
              return undefined;
            }
            return result;
          }
        )
      )
    ),
    ([thought, , action, result]) => {
      const functionCalls = parseFunctionCalls(action);
      return {
        thought,
        action,
        result,
        functionCalls,
      };
    }
  )
);

export function parseActionGroup(expr: string): ActionGroup {
  return expectSingleResult(
    expectEOF(ACTION_GROUP.parse(actionLexer.parse(expr)))
  );
}

enum FunctionTokenKind {
  Identifier,

  TrueLiteral,
  FalseLiteral,
  NumberLiteral,
  SingleQuoteStringLiteral,
  DoubleQuoteStringLiteral,
  BackQuoteStringLiteral,

  LParen,
  RParen,
  Space,
  Comma,
  SemiColon,
}

const functionLexer = buildLexer([
  [true, /^'([^'\\]|\\.)*'/g, FunctionTokenKind.SingleQuoteStringLiteral],
  [true, /^"([^"\\]|\\.)*"/g, FunctionTokenKind.DoubleQuoteStringLiteral],
  [true, /^`([^`\\]|\\.)*`/g, FunctionTokenKind.BackQuoteStringLiteral],
  [true, /^[\+\-]?\d+(\.\d+)?/g, FunctionTokenKind.NumberLiteral],
  [true, /^true/g, FunctionTokenKind.TrueLiteral],
  [true, /^false/g, FunctionTokenKind.FalseLiteral],

  [true, /^[a-zA-Z_-][a-zA-Z0-9_-]*/g, FunctionTokenKind.Identifier],

  [true, /^\(/g, FunctionTokenKind.LParen],
  [true, /^\)/g, FunctionTokenKind.RParen],
  [true, /^\,/g, FunctionTokenKind.Comma],
  [true, /^;/g, FunctionTokenKind.SemiColon],
  [false, /^\s+/g, FunctionTokenKind.Space],
]);

const VALUE = rule<FunctionTokenKind, Value>();
VALUE.setPattern(
  p.apply(
    p.alt(
      p.tok(FunctionTokenKind.TrueLiteral),
      p.tok(FunctionTokenKind.FalseLiteral),
      p.tok(FunctionTokenKind.NumberLiteral),
      p.tok(FunctionTokenKind.SingleQuoteStringLiteral),
      p.tok(FunctionTokenKind.DoubleQuoteStringLiteral),
      p.tok(FunctionTokenKind.BackQuoteStringLiteral)
    ),
    (tok) => {
      switch (tok.kind) {
        case FunctionTokenKind.TrueLiteral:
          return { type: "boolean", value: true };
        case FunctionTokenKind.FalseLiteral:
          return { type: "boolean", value: false };
        case FunctionTokenKind.NumberLiteral:
          return { type: "number", value: +tok.text };
        case FunctionTokenKind.SingleQuoteStringLiteral:
          return {
            type: "string",
            value: tok.text
              .substring(1, tok.text.length - 1)
              .replaceAll(`\\'`, `'`),
          };
        case FunctionTokenKind.DoubleQuoteStringLiteral:
          return {
            type: "string",
            value: tok.text
              .substring(1, tok.text.length - 1)
              .replaceAll(`\\"`, `"`),
          };
        case FunctionTokenKind.BackQuoteStringLiteral:
          return {
            type: "string",
            value: tok.text
              .substring(1, tok.text.length - 1)
              .replaceAll("\\`", "`"),
          };
        default:
          const exhaustiveCheck: never = tok;
          throw new Error(
            `Unsupported value kind: ${(exhaustiveCheck as any).text}`
          );
      }
    }
  )
);

const CALL = rule<FunctionTokenKind, Call>();
CALL.setPattern(
  p.apply(
    p.seq(
      p.tok(FunctionTokenKind.Identifier),
      p.tok(FunctionTokenKind.LParen),
      p.opt_sc(p.list(p.alt(VALUE, CALL), p.tok(FunctionTokenKind.Comma))),
      p.tok(FunctionTokenKind.RParen)
    ),
    ([{ text: name }, , maybeArgs]) => ({
      type: "call",
      name,
      args: maybeArgs ?? [],
    })
  )
);

const CALLS = rule<FunctionTokenKind, Call[]>();
CALLS.setPattern(
  p.apply(
    p.kleft(
      p.list(
        CALL,
        p.seq(
          p.tok(FunctionTokenKind.SemiColon),
          p.rep_sc(p.tok(FunctionTokenKind.SemiColon))
        )
      ),
      p.rep_sc(p.tok(FunctionTokenKind.SemiColon))
    ),
    (calls) => calls
  )
);

export function parseFunctionCall(expr: string): Call {
  return expectSingleResult(expectEOF(CALL.parse(functionLexer.parse(expr))));
}

export function parseFunctionCalls(expr: string): Call[] {
  return expectSingleResult(expectEOF(CALLS.parse(functionLexer.parse(expr))));
}
