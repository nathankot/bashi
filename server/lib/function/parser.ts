import * as p from "typescript-parsec";
import { parseDate } from "chrono";

import {
  Token,
  buildLexer,
  rule,
  expectEOF,
  expectSingleResult,
} from "typescript-parsec";

import { LogFn } from "@lib/log.ts";

import argumentParsers from "./argumentParsers.ts";

import {
  FunctionCall,
  FunctionSet,
  FunctionDefinition,
  Argument,
} from "./types.ts";

enum TokenKind {
  Identifier,

  TrueLiteral,
  FalseLiteral,
  NumberLiteral,
  SingleQuoteStringLiteral,
  DoubleQuoteStringLiteral,
  BackQuoteStringLiteral,

  LParen,
  RParen,
  Space,
  Comma,
}

const lexer = buildLexer([
  [true, /^'([^'\\]|\\.)*'/g, TokenKind.SingleQuoteStringLiteral],
  [true, /^"([^"\\]|\\.)*"/g, TokenKind.DoubleQuoteStringLiteral],
  [true, /^`([^`\\]|\\.)*`/g, TokenKind.BackQuoteStringLiteral],
  [true, /^[\+\-]?\d+(\.\d+)?/g, TokenKind.NumberLiteral],
  [true, /^true/g, TokenKind.TrueLiteral],
  [true, /^false/g, TokenKind.FalseLiteral],

  [true, /^[a-zA-Z_-][a-zA-Z0-9_-]*/g, TokenKind.Identifier],

  [true, /^\(/g, TokenKind.LParen],
  [true, /^\)/g, TokenKind.RParen],
  [true, /^\,/g, TokenKind.Comma],
  [false, /^\s+/g, TokenKind.Space],
]);

function applyLiteral(
  tok:
    | Token<TokenKind.TrueLiteral>
    | Token<TokenKind.FalseLiteral>
    | Token<TokenKind.NumberLiteral>
    | Token<TokenKind.SingleQuoteStringLiteral>
    | Token<TokenKind.DoubleQuoteStringLiteral>
    | Token<TokenKind.BackQuoteStringLiteral>
): Argument {
  switch (tok.kind) {
    case TokenKind.TrueLiteral:
      return true;
    case TokenKind.FalseLiteral:
      return false;
    case TokenKind.NumberLiteral:
      return +tok.text;
    case TokenKind.SingleQuoteStringLiteral:
      return tok.text.substring(1, tok.text.length - 1).replaceAll(`\\'`, `'`);
    case TokenKind.DoubleQuoteStringLiteral:
      return tok.text.substring(1, tok.text.length - 1).replaceAll(`\\"`, `"`);
    case TokenKind.BackQuoteStringLiteral:
      return tok.text.substring(1, tok.text.length - 1).replaceAll("\\`", "`");
    default:
      const exhaustiveCheck: never = tok;
      throw new Error(
        `Unsupported argument kind: ${(exhaustiveCheck as any).text}`
      );
  }
}

function applyCall(
  toks: [
    Token<TokenKind.Identifier>,
    Token<TokenKind.LParen>,
    undefined | Argument[],
    Token<TokenKind.RParen>
  ]
): FunctionCall & { type: "parsed" } {
  const [{ text: name }, , maybeArgs] = toks;
  return {
    type: "parsed",
    line: "",
    name,
    args: maybeArgs ?? [],
  };
}

const ARG = rule<TokenKind, Argument>();
ARG.setPattern(
  p.apply(
    p.alt(
      p.tok(TokenKind.TrueLiteral),
      p.tok(TokenKind.FalseLiteral),
      p.tok(TokenKind.NumberLiteral),
      p.tok(TokenKind.SingleQuoteStringLiteral),
      p.tok(TokenKind.DoubleQuoteStringLiteral),
      p.tok(TokenKind.BackQuoteStringLiteral)
    ),
    applyLiteral
  )
);

const CALL = rule<TokenKind, FunctionCall & { type: "parsed" }>();
CALL.setPattern(
  p.apply(
    p.seq(
      p.tok(TokenKind.Identifier),
      p.tok(TokenKind.LParen),
      p.opt(p.list(ARG, p.tok(TokenKind.Comma))),
      p.tok(TokenKind.RParen)
    ),
    applyCall
  )
);

export function evaluate(expr: string): FunctionCall & { type: "parsed" } {
  return expectSingleResult(expectEOF(CALL.parse(lexer.parse(expr))));
}

export function parseFromModelResult(
  {
    log,
    now,
    knownFunctions,
  }: {
    log: LogFn;
    now: Date;
    knownFunctions: FunctionSet;
  },
  text: string
): FunctionCall[] {
  let result: FunctionCall[] = [];

  for (const line of text.split("\n")) {
    if (line.trim() === "") {
      continue;
    }
    if (line.length < 3) {
      // min number of chars for a valid function is 3
      continue;
    }
    if (line === "```") {
      continue;
    }
    try {
      const parsed = {
        ...evaluate(line),
        line,
      };
      const knownFunction = knownFunctions[parsed.name];
      // Check that the function is known
      if (knownFunction == null) {
        result.push({
          ...parsed,
          type: "invalid",
          invalid_reason: "unknown_function",
        });
        continue;
      }
      if (!checkArgumentsValid(knownFunction, parsed.args)) {
        result.push({
          ...parsed,
          type: "invalid",
          invalid_reason: "invalid_arguments",
        });
        continue;
      }

      // Do any additional argument parsing:
      parsed.argsParsed = knownFunction.args.map((argDef, i) =>
        (argDef.parse ?? []).reduce((a, e) => {
          const rawValue = parsed.args[i];
          if (rawValue == null) {
            return a;
          }
          try {
            const argParser = argumentParsers[e];
            switch (argParser.inputType) {
              case "string":
                if (typeof rawValue !== "string") {
                  throw new Error(
                    `expected parser input to be string got ${typeof rawValue}`
                  );
                }
                return {
                  ...a,
                  [e]: argParser.fn(
                    { now, chronoParseDate: parseDate },
                    rawValue
                  ),
                };

              default:
                const exhaustiveCheck: never = argParser.inputType;
                throw new Error(
                  `unexpected parser input type: ${exhaustiveCheck}`
                );
            }
          } catch (e) {
            log("error", e);
            return a;
          }
        }, {})
      );

      result.push(parsed);
    } catch (e) {
      result.push({
        type: "parse_error",
        line,
        error: (e as any).message ?? "",
      });
    }
  }

  return result;
}

export function checkArgumentsValid(
  knownFunction: FunctionDefinition,
  args: Argument[]
): boolean {
  if (knownFunction.args.length !== args.length) {
    return false;
  }
  for (let i = 0; i < knownFunction.args.length; i++) {
    const argDef = knownFunction.args[i]!;
    const arg = args[i]!;
    if (typeof arg !== argDef.type) {
      return false;
    }
  }
  return true;
}
