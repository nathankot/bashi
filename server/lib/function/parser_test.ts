import { assertSnapshot } from "std/testing/snapshot.ts";
import { evaluate } from "./parser.ts";

for (const expr of [
  "someCall()",
  "someCall('a', 111)",
  "someCall('a \\'ha\\'', 111)",
  "someCall(`a`, 111)",
  "someCall(`a \\`ha\\``, 111)",
  `someCall("a", 111)`,
  `someCall("a \\"ha\\"", 111)`,
  `someCall("a", 111,true)`,
  `someCall("a", 111, false)`,
  `someCall("a false true 123 b", 111, false)`,
  `someCall(-123)`,
  `some_call()`,
  `some-call()`,
  `SOMECALL()`,
  `malformed("a"`,
  `someCall(-123.500)`,
  "testing blah",
  "",
]) {
  Deno.test(expr === "" ? "empty string" : expr, (t) => {
    try {
      const result = evaluate(expr);
      assertSnapshot(t, result);
    } catch (e) {
      assertSnapshot(t, (e as Error).message);
    }
  });
}
