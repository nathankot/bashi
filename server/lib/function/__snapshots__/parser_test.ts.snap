export const snapshot = {};

snapshot[`someCall() 1`] = `
{
  args: [
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall('a', 111) 1`] = `
{
  args: [
    "a",
    111,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall('a \\'ha\\'', 111) 1`] = `
{
  args: [
    "a 'ha'",
    111,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall(\`a\`, 111) 1`] = `
{
  args: [
    "a",
    111,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall(\`a \\\`ha\\\`\`, 111) 1`] = `
{
  args: [
    "a \`ha\`",
    111,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall("a", 111) 1`] = `
{
  args: [
    "a",
    111,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall("a \\"ha\\"", 111) 1`] = `
{
  args: [
    'a "ha"',
    111,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall("a", 111,true) 1`] = `
{
  args: [
    "a",
    111,
    true,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall("a", 111, false) 1`] = `
{
  args: [
    "a",
    111,
    false,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall("a false true 123 b", 111, false) 1`] = `
{
  args: [
    "a false true 123 b",
    111,
    false,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`someCall(-123) 1`] = `
{
  args: [
    -123,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`some_call() 1`] = `
{
  args: [
  ],
  line: "",
  name: "some_call",
  type: "parsed",
}
`;

snapshot[`some-call() 1`] = `
{
  args: [
  ],
  line: "",
  name: "some-call",
  type: "parsed",
}
`;

snapshot[`SOMECALL() 1`] = `
{
  args: [
  ],
  line: "",
  name: "SOMECALL",
  type: "parsed",
}
`;

snapshot[`malformed("a" 1`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;

snapshot[`someCall(-123.500) 1`] = `
{
  args: [
    -123.5,
  ],
  line: "",
  name: "someCall",
  type: "parsed",
}
`;

snapshot[`testing blah 1`] = `'{"index":8,"rowBegin":1,"columnBegin":9,"rowEnd":1,"columnEnd":13}: Unable to consume token: blah'`;

snapshot[`empty string 1`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;
